### PR TITLE
imxrt:serial rxdma reporting incorrect characters

### PR DESCRIPTION
**Pull request by danielappiagyei-bc**
## Summary
Fixes issue https://github.com/apache/nuttx/issues/10912
## Impact
TODO
## Testing
TODO

# **DO NOT MERGE YET** #

### CODE REVIEW

File: arch/arm/src/imxrt/imxrt_serial.c [hunk 1]
```diff
@@ -60,6 +60,24 @@
 
 #ifdef USE_SERIALDRIVER
 
+#if defined(SERIAL_HAVE_TXDMA)
```
Comment (xiaoxiang781216) [inline arch/arm/src/imxrt/imxrt_serial.c:63]:
```suggestion
#if defined(SERIAL_HAVE_TXDMA) && CONFIG_IMXRT_EDMA_NTCD < 2
#  error Each LPUART using TXDMA requires 2 EDMA TCDs for setting up transfers
#endif
```

File: arch/arm/src/imxrt/imxrt_serial.c [hunk 2]
```diff
@@ -60,6 +60,24 @@
...
+#if CONFIG_IMXRT_EDMA_NTCD < 2
+#error Each LPUART using TXDMA requires 2 EDMA TCDs for setting up transfers
+#endif
+#endif /* SERIAL_HAVE_TXDMA */
+
+#if defined(SERIAL_HAVE_RXDMA)
+#if CONFIG_IMXRT_EDMA_NTCD < 1
+#error Each LPUART using RXDMA requires 1 EDMA TCD for setting up transfers
```
Comment (xiaoxiang781216) [inline arch/arm/src/imxrt/imxrt_serial.c:71]: ditto

File: arch/arm/src/imxrt/imxrt_serial.c [hunk 3]
```diff
@@ -60,6 +60,24 @@
...
+#endif
+#endif /* SERIAL_HAVE_RXDMA */
+
+#if defined(SERIAL_HAVE_TXDMA) && defined(SERIAL_HAVE_RXDMA) 
+#if CONFIG_IMXRT_EDMA_NTCD < 3
```
Comment (xiaoxiang781216) [inline arch/arm/src/imxrt/imxrt_serial.c:76]: ditto

File: arch/arm/src/imxrt/imxrt_serial.c [hunk 4]
```diff
@@ -1531,12 +1560,13 @@ static int imxrt_dma_setup(struct uart_dev_s *dev)
       modifyreg32(priv->uartbase + IMXRT_LPUART_BAUD_OFFSET,
                   0, LPUART_BAUD_RDMAE);
 
-      /* Enable itnerrupt on Idel and erros */
+      /* Enable interrupt on Idle and errors */
```
Comment (davids5) [inline arch/arm/src/imxrt/imxrt_serial.c:1563]: Add Over run to Comment

File: arch/arm/src/imxrt/imxrt_serial.c [hunk 5]
```diff
@@ -1370,11 +1388,22 @@ static inline void imxrt_serialout(struct imxrt_uart_s *priv,
  ****************************************************************************/
 
 #ifdef SERIAL_HAVE_RXDMA
-static int imxrt_dma_nextrx(struct imxrt_uart_s *priv)
+static unsigned int imxrt_dma_nextrx(struct imxrt_uart_s *priv)
```
Comment (davids5) [inline arch/arm/src/imxrt/imxrt_serial.c:1391]: :1

File: arch/arm/src/imxrt/imxrt_serial.c (hunk 1 above)
Comment (davids5) [inline arch/arm/src/imxrt/imxrt_serial.c:63]: the TCD pool is shared across all devices.  So this is only a Band-Aid. For the minimal case

**Review by davids5 — CHANGES_REQUESTED** [review]
@danielappiagyei-bc Great! - The residual change needs to be applied the S32K, and K66

File: arch/arm/src/imxrt/imxrt_serial.c (hunk 1 above)
Comment (danielappiagyei-bc) [reply arch/arm/src/imxrt/imxrt_serial.c]: done!

File: arch/arm/src/imxrt/imxrt_serial.c (hunk 4 above)
Comment (danielappiagyei-bc) [reply arch/arm/src/imxrt/imxrt_serial.c]: done!

File: arch/arm/src/imxrt/imxrt_serial.c (hunk 1 above)
Comment (danielappiagyei-bc) [reply arch/arm/src/imxrt/imxrt_serial.c]: totally agree. I just found myself using a single lpuart device with DMA for the first time and it wasn't obvious to me why dma wasn't working properly (no transfers were occurring) until I stumbled upon the fact that you need to set `CONFIG_IMXRT_EDMA_NTCD` appropriately. I wasn't sure of a better way to communicate this to the user and was hoping these compile-time checks would help at least a few people :)

**Comment by danielappiagyei-bc** [comment]
> @danielappiagyei-bc Great! - The residual change needs to be applied the S32K, and K66

Hi @davids5 , I've now applied these changes to the S32K devices but couldn't find a "k66", could you point me to the file? 
Also I didn't enable an interrupt for overruns on these S32K devices, would you like me to? Thanks

**Comment by davids5** [comment]
> > @danielappiagyei-bc Great! - The residual change needs to be applied the S32K, and K66
> 
> Hi @davids5 , I've now applied these changes to the S32K devices but couldn't find a "k66", could you point me to the file? Also I didn't enable an interrupt for overruns on these S32K devices, would you like me to? Thanks
Sorry I called it by name not family https://github.com/apache/nuttx/blob/master/arch/arm/src/kinetis/kinetis_edma.c

File: arch/arm/src/imxrt/Kconfig [hunk 6]
```diff
@@ -183,7 +183,7 @@ config IMXRT_APP_FORMAT_MCUBOOT
 	select IMXRT_HAVE_OTA_PARTITION
 	depends on EXPERIMENTAL
 	---help---
-		The MCUboot support of loading the firmware images.		
+		The MCUboot support of loading the firmware images.
```
Comment (pkarashchenko) [inline arch/arm/src/imxrt/Kconfig:186]:
```suggestion
		The MCUboot support of loading the firmware images.
```

Comment (danielappiagyei-bc) [reply arch/arm/src/imxrt/Kconfig]: done, good catch :)

**Comment by davids5** [comment]
> > > @danielappiagyei-bc Great! - The residual change needs to be applied the S32K, and K66
> > 
> > 
> > Hi @davids5 , I've now applied these changes to the S32K devices but couldn't find a "k66", could you point me to the file? Also I didn't enable an interrupt for overruns on these S32K devices, would you like me to? Thanks
> 
> Sorry I called it by name not family https://github.com/apache/nuttx/blob/master/arch/arm/src/kinetis/kinetis_edma.c

So strike this.  See https://github.com/apache/nuttx/issues/10912#issuecomment-1777212451

You might want to rebase on master after https://github.com/apache/nuttx/pull/11020 comes in and bring in all the other bits not related to serial or DMA

**Comment by danielappiagyei-bc** [comment]
> So strike this. See [#10912 (comment)](https://github.com/apache/nuttx/issues/10912#issuecomment-1777212451)
> 
> You might want to rebase on master after #11020 comes in and bring in all the other bits not related to serial or DMA

Thanks david, I will close this PR and create a new one for the leftover changes. This will be:
- typos
- adding `IMXRT_SERIAL_RXDMA_BUFFER_SIZE` as a kconfig
- add missing `uintptr_t base;` declaration to `arch/arm/src/imxrt/imxrt_edma.c` so compiling with `CONFIG_DEBUG_DMA=y` works
- enabling `LPUART_CTRL_ORIE` for overruns in `arch/arm/src/imxrt/imxrt_serial.c`. It's an error condition just like the noise, framing, and parity error interrupts we have enabled and we are already explicitly clearing it in [here](https://github.com/apache/nuttx/blob/master/arch/arm/src/imxrt/imxrt_serial.c#L1792)

**Comment by danielappiagyei-bc** [comment]
> So strike this. See https://github.com/apache/nuttx/issues/10912#issuecomment-1777212451

> You might want to rebase on master after https://github.com/apache/nuttx/pull/11020 comes in and bring in all the other bits not related to serial or DMA

Per david, he has found the cause to the issue and is fixing in his PR. See https://github.com/apache/nuttx/pull/10911#issuecomment-1777940825 for what I will add in followup PR

**Pull request closed without merging**